app/models/action.rb

```diff
@@ -2,18 +2,40 @@ class Action < ApplicationRecord
   has_paper_trail versions: { class_name: 'Version::ActionVersion' }
 
   belongs_to :user
-  belongs_to :contact
+  belongs_to :contact, optional: true
+  has_many :subactions, class_name: 'Action',
+                        foreign_key: 'bulk_action_id',
+                        inverse_of: :bulk_action,
+                        dependent: :destroy
+  belongs_to :bulk_action, class_name: 'Action', optional: true
 
   validates :operation, inclusion: { in: proc { |action| action.class.valid_operations } }
 
   class << self
     def valid_operations
-      %w[update]
+      %w[update bulk_update]
     end
   end
 
   def notification_key
-    raise 'Action object is missing' unless contact
+    raise 'Action object is missing' unless bulk_action? || contact
+
     "contact_#{operation}".to_sym
   end
+
+  def bulk_action?
+    !!subactions.exists?
+  end
+
+  def to_non_available_contact_codes
+    return [] unless bulk_action?
+
+    subactions.map do |a|
+      {
+        code: a.contact&.code,
```

Comment:
but if a.contact is nil, then code will equal nil. What is the point of element of subactions list if it will store some element with nil value? Are you sure is it need this logic here?

Reply:
because contact is now optional, but in this context, yes, it cannot be nil, because we need code attribute